cmd/imagec/docker.go

```diff
@@ -181,47 +183,64 @@ func FetchImageBlob(options ImageCOptions, image *ImageWithMeta) error {
 	}()
 
 	// Open the file so that we can use it as a io.Reader for sha256 calculation
-	in, err := os.Open(string(imageFileName))
+	imageFile, err := os.Open(string(imageFileName))
 	if err != nil {
-		return err
+		return diffID, err
 	}
-	defer in.Close()
+	defer imageFile.Close()
 
-	h := sha256.New()
+	// blobSum is the sha of the compressed layer
+	blobSum := sha256.New()
+
+	// diffIDSum is the sha of the uncompressed layer
+	diffIDSum := sha256.New()
+
+	// blobTr is an io.TeeReader that writes bytes to blobSum that it reads from imageFile
+	// see https://golang.org/pkg/io/#TeeReader
+	blobTr := io.TeeReader(imageFile, blobSum)
 
 	progress.Update(po, image.String(), "Verifying Checksum")
-	if _, cerr := io.Copy(h, in); cerr != nil {
-		return cerr
+	tar, err := archive.DecompressStream(blobTr)
+	if err != nil {
+		return diffID, err
 	}
 
-	// Calculate the sum
-	sum := fmt.Sprintf("sha256:%x", h.Sum(nil))
-	if sum != layer {
-		return fmt.Errorf("Failed to validate layer checksum. Expected %s got %s", layer, sum)
+	// Copy bytes from decompressed layer into diffIDSum to calculate diffID
+	if _, cerr := io.Copy(diffIDSum, tar); cerr != nil {
+		return diffID, cerr
 	}
 
+	bs := fmt.Sprintf("sha256:%x", blobSum.Sum(nil))
+	if bs != layer {
```

Comment:
Yeah, it's doing a sum over the compressed layer. So we do the same and just check it against the one in the manifest to make sure they match.

This 1.10 design doc was an enlightening read for me, especially this part:  https://gist.github.com/aaronlehmann/b42a2eaf633fc949f93b#id-definitions-and-calculations